marimo/_save/save.py

```diff
@@ -511,3 +364,289 @@ def __exit__(
             tmpio.seek(0)
             write_traceback(tmpio.read())
         return False
+
+
+@singledispatch
+def _cache_invocation(
+    arg: Any,
+    loader: Union[LoaderPartial, Loader, LoaderType],
+    *args: Any,
+    frame_offset: int = 1,
+    **kwargs: Any,
+) -> Union[_cache_call, _cache_context]:
+    del loader, args, kwargs, frame_offset
+    raise TypeError(f"Invalid type for cache: {type(arg)}")
+
+
+def _invoke_call(
+    _fn: Callable[..., Any] | None,
+    loader: Union[LoaderPartial, Loader, LoaderType],
+    *args: Any,
+    frame_offset: int = 1,
+    **kwargs: Any,
+) -> _cache_call:
+    if isinstance(loader, Loader):
+        raise TypeError(
+            "A loader instance cannot be passed to cache directly. "
+            f"Specify a loader type (e.g. `{loader.__class__}`) or a loader "
+            f"partial (e.g. `{loader.__class__}.partial(arg=value)`)."
+        )
+    elif isinstance(loader, type) and issubclass(loader, Loader):
+        loader = cast(Loader, loader).partial()
+
+    if not isinstance(loader, LoaderPartial):
+        raise TypeError(
+            "Invalid loader type. "
+            f"Expected a loader partial, got {type(loader)}."
+        )
+    return _cache_call(
+        _fn, loader, *args, frame_offset=frame_offset + 1, **kwargs
+    )
+
+
+@_cache_invocation.register
+def _invoke_call_none(
+    _fn: None,
+    loader: Union[LoaderPartial, Loader, LoaderType],
+    *args: Any,
+    frame_offset: int = 1,
+    **kwargs: Any,
+) -> _cache_call:
+    return _invoke_call(
+        _fn, loader, *args, frame_offset=frame_offset + 1, **kwargs
+    )
+
+
+@_cache_invocation.register
+def _invoke_call_fn(
+    # mypy would like some generics, but this breaks the singledispatch
+    _fn: abc.Callable,  # type: ignore[type-arg]
+    loader: Union[LoaderPartial, Loader, LoaderType],
+    *args: Any,
+    frame_offset: int = 1,
+    **kwargs: Any,
+) -> _cache_call:
+    return _invoke_call(
+        _fn, loader, *args, frame_offset=frame_offset + 1, **kwargs
+    )
+
+
+@_cache_invocation.register
+def _invoke_context(
+    name: str,
+    loader: Union[LoaderPartial, Loader, LoaderType],
+    *args: Any,
+    frame_offset: int = 1,
+    **kwargs: Any,
+) -> _cache_context:
+    del frame_offset
+
+    if isinstance(loader, LoaderPartial):
+        loader = loader.create_or_reconfigure(name)()
+    elif isinstance(loader, type) and issubclass(loader, Loader):
+        # Create through partial for meaningful error message.
+        loader = cast(Loader, loader).partial()
+        assert isinstance(loader, LoaderPartial), (
+            UNEXPECTED_FAILURE_BOILERPLATE
+        )
+        loader = loader.create_or_reconfigure(name)()
+    return _cache_context(name, loader, *args, **kwargs)
+
+
+def cache(
+    name: Union[str, Optional[Callable[..., Any]]] = None,
+    *args: Any,
```

Comment:
Why change the signature to use varargs/varkwargs?

Reply:
IIUC this can dispatch to either `_cache_call` or `_cache_context`, but they take different args/kwargs, so variability is needed. Rather than using `Any`, tho, I think this could have more specific type annotations using [unpack](https://typing.readthedocs.io/en/latest/spec/callables.html#unpack-kwargs) and a union those two signatures' args/kwargs?

Reply:
`Any` is pretty consistent with other uses of `*` packing in the codebase, but I did leave a comment, since `*` is more to keep values ambiguous until they're fed into the correct constructors

Reply:
Yea, to be clear, wasn't saying that needed to happen, just that more specific typing was an option if desired -- while a nice addition, `unpack` is pretty recent and not super widely used (in marimo and beyond, ime)

Thanks for your work on this! I'm trying this branch out on a couple long computations rn 🚀